### PR TITLE
Nuke empty communities from db

### PR DESCRIPTION
**Pull request by Matombo**
A script that helps keeping the homeserver clean from empty communities.
Related to issue #3023.

Developer Certificate of Origin
Version 1.1
Signed-off-by: Werner Sembach <werner.sembach at fau.de>

### CODE REVIEW

File: scripts-dev/nuke-empty-communities-from-db.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    cur = db.cursor()
+elif argv[1] == "postgresql":
+    db = psycopg2.connect('host=' + argv[2] + ' dbname=' + argv[3] + ' user=' + argv[4] + ' password=' + argv[5])
+    cur = db.cursor()
+else:
+    print "This should be unreachable, report a bug."
+    quit()
+
+with db:
+    cur.execute('SELECT group_id FROM groups;')
```
Comment (krombel) [inline scripts-dev/nuke-empty-communities-from-db.py]: You can extend this query to the following to not need the check in the `for` loop:
```
SELECT g.group_id FROM groups g
    WHERE (SELECT count(*) from group_users u WHERE g.group_id = u.group_id) = 0
```

**Comment by ArchangeGabriel** [comment]
Does it works when using PostgreSQL through UNIX socket? I believe not, because you are forcing the use of too many parameters for this case.

**Comment by Matombo** [comment]
I don't have much PostgreSQL knowledge.
Can you plese point me out on which parameters of those 4 are necessary and which are optional?

**Comment by ArchangeGabriel** [comment]
AFAIK only dbname is required, because you can use `peer` authentication when using UNIX Socket, so that no username, password or host is required.

File: scripts-dev/nuke-empty-communities-from-db.py [hunk 2]
```diff
@@ -1,33 +1,49 @@
 #!/usr/bin/env python2
 
 from sys import argv
+from getopt import getopt
 import sqlite3
 import psycopg2
 
 if len(argv) < 3 or (argv[1] != "sqlite" and argv[1] != "postgresql") or (argv[1] == "postgresql" and len(argv) < 6):
```
Comment (ArchangeGabriel) [inline scripts-dev/nuke-empty-communities-from-db.py]: I think the `6` add the end of this line still makes options mandatory in practice.

**Comment by Matombo** [comment]
Woops forget that line
Corrected
Thanks

**Comment by Matombo** [comment]
What exactly is the Synapse Sytest?
Do i have to do anything for it to finish?

**Comment by Matombo** [comment]
should i move this script to contrib folder?

**Comment by dragetd** [comment]
Wouldn't it be a lot better if we stick to prepared states instead of executing MySQL Strings?

**Comment by aaronraimist** [comment]
There is an API to delete groups now so this script probably isn't needed anymore https://github.com/matrix-org/synapse/blob/master/docs/admin_api/delete_group.md

**Comment by Matombo** [comment]
The api only deletes single groups by id. This script scans the database for empty groups and deletes them.

This script still serves an unique function but needs to be updated to use the new api.

**Comment by Matombo** [comment]
Thinking about the new api  this would actually limit the script funtionality.

The api uses a http interface, so it can only be used while the server is running. This script can be directly executed on the database of an online or offline instance.

Also, the api currently only supports deleting groups, not to search all groups and select them according to members count. The script still needs to issue sql statements even when using the api.

**Comment by Matombo** [comment]
Made obsolete by: https://github.com/matrix-org/synapse/pull/6453

**Pull request closed without merging**